Feature elastic kws

miahunsicker opened the pull request:
- Ability to sort only on award_id
- Need to add ES_HOSTNAME to env, please see me for endpoint. 
- Update requirements for pandas, and add elasticsearch python requirement

tony-sappe commented on requirements/requirements.txt:
```diff
@@ -17,12 +17,13 @@ Django==1.11.4
 Markdown==2.6.6
 model-mommy==1.3.0
 openpyxl==2.4.7
-pandas==0.18.1
```
Let's not change this right now

tony-sappe commented on usaspending_api/search/v2/elasticsearch_helper.py:
```diff
@@ -0,0 +1,70 @@
+TRANSACTIONS_INDEX_ROOT = settings.TRANSACTIONS_INDEX_ROOT
+
+
+CLIENT = Elasticsearch(ES_HOSTNAME)
```
This is OK for now. But it would be better to store this ES client create in a more centralized part of the codebase for any future endpoints to use. It could be a context manager to provide easier validation and error-messages

tony-sappe commented on usaspending_api/search/v2/elasticsearch_helper.py:
```diff
@@ -0,0 +1,70 @@
+    query_fields = [TRANSACTIONS_LOOKUP[i] for i in fields]
+    query_sort = TRANSACTIONS_LOOKUP[sort]
+    query = {
+        '_source' : query_fields,
```
PEP8 issues (here and other files)

tony-sappe commented on usaspending_api/search/v2/views/search.py:
```diff
@@ -735,6 +737,70 @@ def post(self, request):
+    
+        if order not in ["asc", "desc"]:
+            raise InvalidParameterException("Invalid value for order: {}".format(order))
+        sort = json_request.get("sort", fields[0])
```
could this be locate with the other request parsers?

tony-sappe commented on usaspending_api/search/v2/views/search.py:
```diff
@@ -735,6 +737,70 @@ def post(self, request):
+        
+        lower_limit = (page - 1) * limit
+        upper_limit = page * limit
+        transaction_types = ["Contracts", "Grants", "Direct Payments", "Loans", "Other"]
```
is this used?

tony-sappe commented on usaspending_api/search/v2/views/search.py:
```diff
@@ -735,6 +737,70 @@ def post(self, request):
+        
+        #fields.append('awarding_agency_id')
+
+        queryset, total = search_transactions(filters, fields, sort, 
```
It might be better to not call this a queryset. mostly due to the fact that all other querysets are Django Querysets

miahunsicker commented:
This pr is for keyword search, spending by transaction endpoint, `/api/v2/search/spending_by_transaction/`. This is an integration with Elasticsearch to deliver faster results for keyword search. 
- Added python requirements to query elasticsearch. 
- Added view
- Requires new env variable for elastic host endpoint
- Added lookups and scripts to handle queries. 
https://github.com/fedspendingtransparency/usaspending-api/wiki#spending-by-transaction

- [ ] Frontend Review  @ebdabbs 
- [ ] Backend Review @tony-sappe 
- [ ] Ops Review/Tasks  @hsharma96

- TODO: ES tests.

tony-sappe commented:
Pagenation Issue:

Using this payload:
`{
  "filters": {
       "keyword": "Booz Allen",
       "award_type_codes":["A","B","C","D"]
       },
  "fields": ["Award ID", "Recipient Name", "Action Date", "Transaction Amount", "Awarding Agency", "Awarding Sub Agency", "Award Type", "Funding Agency", "Funding Sub Agency", "Mod"],
  "sort": "Funding Agency",
  "order": "asc",
  "page": 167,
  "limit": 60
}`

Page metadata:
`    "page_metadata": {
        "page": 166,
        "hasNext": true
    }
}`

However, requesting page 167 results in an ES error

tony-sappe commented:
Known issues with pagination (possibly when the last result aligns exactly on a page) Will be examined in a later sprint.
currently only 3 FYs of data in Elasticsearch

tony-sappe commented:
Out of these fields:  ["Award ID", "Recipient Name", "Action Date", "Transaction Amount", "Awarding Agency", "Awarding Sub Agency", "Award Type", "Funding Agency", "Funding Sub Agency", "Mod"]

These do not work as the sort key:
- Mod
- Awarding Sub Agency
